Namespace optimization for hdfs data segments.

akashdw opened the pull request:
This PR is an optimization to use less namespace to store segments into hdfs, this is achieved by  eliminating shard number directory from the current directory structure.

Current directory structure is :
`hdfs://nn1/hdfs_base_directory/data_source_name/interval/version/shardNum/index.zip`
After this change, structure will be:
`hdfs://nn1/hdfs_base_directory/data_source_name/interval/version/shardNum_index.zip`
and 
`hdfs://nn1/hdfs_base_directory/data_source_name/interval/version/shardNum_descriptor.json`

Tables with high number of shards for a version will see approximately ~33% reduction in namespace usage.

pjain1 commented:
does the insert segment tool still works after this changes ? there are various references to "index.zip" and "descriptor.json" file name in the documentation and unit tests. Is there any change required there ?

erikdubbelboer commented:
This will break all current hdfs segments that people have. If someone for example issues an index task with current segments as input it will fail because it won't be able to find these segments as it looks for them in the new path while they are in the old path.

akashdw commented:
@erikdubbelboer I don't think index task will look for segments in the new path. Old segments metadata(`loadSpec.path`) still be pointing to old path and any indexing for those segments can still be done. This PR only changes the directory structure for newly **created** segments and setting the loadSpec with new directory structure.

himanshug commented:
also, HdfsDataSegmentKiller assumes that directory structure and will need to be updated to work with both old and new directory structure.

akashdw commented:
@pjain1 @himanshug @gianm Thanks for reviewing this PR. I made changes in HdfsDataSegmentFinder, HdfsDataSegmentFKiller to understand new storage path as well as old path and updated the doc. Some of the tests in `druid-indexing-hadoop` project are still using old path format because those tests uses `file` scheme to store data segments and this PR does not change the current behavior for `file` scheme.

himanshug commented on extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java:
```diff
@@ -99,13 +118,20 @@ public void killAll() throws IOException
+    }
     try {
-      return fs.delete(path, false);
+      if(fs.listStatus(path).length == 0)
```
can this be `if (fs.listStatus(path).length == 0 && fs.delete(path,false)) ` so that delete call return status is not ignored

himanshug commented on extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java:
```diff
@@ -99,13 +118,20 @@ public void killAll() throws IOException
-  private boolean safeNonRecursiveDelete(FileSystem fs, Path path)
+  private void deleteEmptyParents(FileSystem fs, Path path)
   {
+    if(path.equals(storageDirectory)) {
```
this does not work if storageDirectory has changed since the time segment, being deleted, was created.

himanshug replied:
also does it work if only out of of storageDirectory or path are fully qualified with NN details?

akashdw replied:
Yup, it won't work if storageDirectory has changed and also there is no way to figure out old storage directory. 
changing it to do comparison on qualified paths.

himanshug commented on extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPusher.java:
```diff
@@ -107,18 +108,23 @@ public DataSegment push(File inDir, DataSegment segment) throws IOException
           storageDir
       ));
+      StringBuilder outFileUriStringBuilder= new StringBuilder(outFile.toUri().toString());
+      outFileUriStringBuilder.setCharAt(outFileUriStringBuilder.lastIndexOf("/"), '_');  //replaces last `/` with `_`
```
can we instead add methods in DataSegmentPusherUtil so that it gives the segment location upto version and partition number is added here appropriately instead of doing the string replace etc

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -413,13 +413,17 @@ public long push() throws IOException
       case "hdfs":
       case "viewfs":
+        StringBuilder indexUriStringBuilder= new StringBuilder(indexOutURI.toString());
+        indexUriStringBuilder.setCharAt(indexUriStringBuilder.lastIndexOf("/"), '_');  //replaces last `/` with `_`
```
same comment here... shouldn't need replacing

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -463,10 +467,18 @@ public long push() throws IOException
+    if("hdfs".equals(outputFS.getScheme()) || "viewfs".equals(outputFS.getScheme()))
+    {
+      StringBuilder descriptorUriStringBuilder= new StringBuilder(finalDescriptorPath.toUri().toString());
+      descriptorUriStringBuilder.setCharAt(descriptorUriStringBuilder.lastIndexOf("/"), '_');  //replaces last `/` with `_`
```
here too

akashdw commented:
Thanks @himanshug. Addressed comments.

nishantmonu51 commented on api/src/main/java/io/druid/segment/loading/DataSegmentPusherUtil.java:
```diff
@@ -64,4 +64,21 @@ public static String getHdfsStorageDir(DataSegment segment)
+   * Due to https://issues.apache.org/jira/browse/HDFS-13 ":" are not allowed in
+   * path names. So we format paths differently for HDFS.
+   */
+  public static String getHdfsStorageDirUptoVersion(DataSegment segment)
```
minor nit : in  getHdfsStorageDir() method above, you can actually reUse this method.

akashdw replied:
👍

akashdw commented:
Thanks @nishantmonu51. Updated DataSegmentPusherUtil.getHdfsStorageDir() to reuse  getHdfsStorageDirUpToVersion().

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -588,6 +608,20 @@ public static Path makeSegmentOutputPath(
     return new Path(prependFSIfNullScheme(fileSystem, basePath), String.format("./%s", segmentDir));
   }
 
+  //This method is very similar to makeSegmentOutputPath() method except it returns output path upto version
```
`makeSegmentOutputPath` appears to be unused now, why not just replace it with this one?

akashdw replied:
`makeSegmentOutputPath` is  public, changing this function might break custom extensions using this method.

akashdw replied:
May be I can mark this one  deprecated ?

gianm replied:
JobHelper isn't in the druid-api module and is not listed on http://druid.io/docs/latest/development/modules.html so I wouldn't consider it part of Druid's public API. So, it's fine to make changes to public methods.

akashdw replied:
👍

gianm commented on api/src/main/java/io/druid/segment/loading/DataSegmentPusherUtil.java:
```diff
@@ -54,14 +54,25 @@ public static String getStorageDir(DataSegment segment)
   public static String getHdfsStorageDir(DataSegment segment)
```
`getHdfsStorageDir` seems unused now, why not replace it with the code from `getHdfsStorageDirUptoVersion`?

gianm replied:
In particular, `getHdfsStorageDir` appears to be used only by `JobHelper.makeSegmentOutputPath`, and _that_ appears to be unused.

akashdw replied:
Same as above, changing this function might break custom extensions using this function.

gianm commented on extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentFinder.java:
```diff
@@ -79,8 +80,16 @@ public HdfsDataSegmentFinder(Configuration config, ObjectMapper mapper)
         final Path path = locatedFileStatus.getPath();
-        if (path.getName().equals("descriptor.json")) {
-          final Path indexZip = new Path(path.getParent(), "index.zip");
+        Path indexZip;
```
indexZip could be moved back into the `if` block and stay final.

gianm commented on extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java:
```diff
@@ -99,13 +118,19 @@ public void killAll() throws IOException
+    }
     try {
-      return fs.delete(path, false);
+      if(fs.listStatus(path).length == 0 && fs.delete(path, false))
```
Please apply codestyle to this file. The formatting is off here.

gianm commented on extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java:
```diff
@@ -99,13 +118,19 @@ public void killAll() throws IOException
-  private boolean safeNonRecursiveDelete(FileSystem fs, Path path)
+  private void deleteEmptyParents(FileSystem fs, Path path)
   {
+    if(fs.makeQualified(path).equals(fs.makeQualified(storageDirectory))) {
```
This check is weak since users can change storageDirectory at any time, and that will affect new segments, but old segments will remain in their old locations and still be killable there. So it's better for single-segment killing logic to avoid depending on storageDirectory in any way.

Maybe use the "version", "interval", and "dataSource" fields of the segment to detect when to stop walking back up the parents?

akashdw replied:
changed it to stop at data source level.

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -384,7 +383,12 @@ public static DataSegment serializeOutIndex(
+    final Path tmpPath = outputFS.getScheme().equals("hdfs") || outputFS.getScheme().equals("viewfs") ? new Path(
+        segmentBasePath,
+        String.format("%d_index.zip.%d", segmentTemplate.getShardSpec().getPartitionNum(), taskAttemptID.getId())
+    ) : new Path(segmentBasePath, String.format("index.zip.%d", taskAttemptID.getId()));
```
This conditional for tmpPath isn't necessary. You could use the "0_index.zip.taskAttemptId" format for all schemes and it would work fine, and be simpler.

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -413,13 +417,18 @@ public long push() throws IOException
 
-    final Path finalIndexZipFilePath = new Path(segmentBasePath, "index.zip");
-    final URI indexOutURI = finalIndexZipFilePath.toUri();
+    Path finalIndexZipFilePath = new Path(segmentBasePath, "index.zip");
```
This isn't very final any more, is it :)

akashdw replied:
he he, changing the variable name :P

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -463,10 +472,21 @@ public long push() throws IOException
     }
+    Path finalDescriptorPath = new Path(segmentBasePath, "descriptor.json");
+
+    if ("hdfs".equals(outputFS.getScheme()) || "viewfs".equals(outputFS.getScheme())) {
```
The proliferation of scheme checks against hdfs/viewfs is, imo, kind of sketchy and confusing. How about doing something like this:

- Add JobHelper.isHdfs that returns `"hdfs".equals(fileSystem.getScheme()) || "viewfs".equals(fileSystem.getScheme())`
- And JobHelper.makeIndexZipPath, and JobHelper.makeDescriptorJsonPath that use JobHelper.isHdfs to make their decisions.
- Remove JobHelper.makeSegmentOutputPath and JobHelper.makeSegmentOutputPathUptoVersionForHdfs, since they wouldn't be needed anymore.

At least then the checks are more contained.

akashdw commented:
Thanks @gianm for the review. Addressed comments. Changed the `HdfsSegmentKiller.deleteEmptyParents()` to stop walking back up the parents at data source level.

himanshug commented:
:+1:

gianm commented on docs/content/development/modules.md:
```diff
@@ -118,7 +118,7 @@ To mark a segment as not used, you need to connect to your metadata storage and
 To start a segment killing task, you need to access the old Coordinator console `http://<COODRINATOR_IP>:<COORDINATOR_PORT/old-console/kill.html` then select the appropriate datasource and then input a time range (e.g. `2000/3000`).
 
-After the killing task ends, both `descriptor.json` and `index.zip` files should be deleted from the data storage.
+After the killing task ends, both `descriptor.json` (`partitionNum_descriptor.json` for `HDFS` data storage)  and `index.zip` (`partitionNum_index.zip` for `HDFS` data storage) files should be deleted from the data storage.
```
HDFS doesn't really need preformatted styling.

gianm commented on extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java:
```diff
@@ -48,42 +48,80 @@ public HdfsDataSegmentKiller(final Configuration config, final HdfsDataSegmentPu
+    final String dataSource = segment.getDataSource();
+    final String segmentPath = getPath(segment).toUri().toString();
+
+    StringBuilder builder = new StringBuilder(segmentPath.split(dataSource)[0]);
```
This isn't robust, what if the segmentPath is `/data/druid/segments/druid/20111001T000000.000Z_20111002T000000.000Z/20111001T000000.000Z/index.zip` because the dataSource name is "druid"? Then this method will return `/data/druid` which isn't right.

One alternative is to have two different deleteEmptyParents logics, one for things named like "index.zip" and one for things like "0_index.zip", using the old safeNonRecursiveDelete method.

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -377,14 +376,16 @@ public static DataSegment serializeOutIndex(
-    final FileSystem outputFS = FileSystem.get(segmentBasePath.toUri(), configuration);
-    final Path tmpPath = new Path(segmentBasePath, String.format("index.zip.%d", taskAttemptID.getId()));
+    final FileSystem outputFS = FileSystem.get(finalIndexZipFilePath.toUri(), configuration);
+
```
nit: one two many newlines here

gianm commented on indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java:
```diff
@@ -576,16 +577,105 @@ private static void createNewZipEntry(ZipOutputStream out, File file) throws IOE
+    return finalIndexZipPath;
+  }
+
+  public static Path makeDescriptorPath(
```
If you want, this and `makeIndexZipPath` could share a lot of logic in a common helper method that just takes the base file name ("index.zip" vs "descriptor.json")

akashdw commented:
Thanks @gianm, addressed comments.

akashdw commented:
@gianm @himanshug Tested these changes in stg environment, puller/pusher/killer looks good. Tested with diff combinations:
 a) data source with segments stored using new format
 b) data source with segments stored using old format
 c) data source with segments stored using mixed format (new + old format).

fjy commented:
@gianm any more comments?

gianm commented:
@fjy looking again now

gianm commented:
@himanshug @nishantmonu51 are you still ok after the recent changes? Go ahead and merge if so.

akashdw commented:
@gianm I was going through the changes and found `maprfs` recently got added in jobHelper. Making a small change here, https://github.com/druid-io/druid/blob/master/indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java#L422 to include `maprfs`.

gianm commented:
👍
